[Dist Autograd] Functional API for Dist Autograd and Dist Optimizer

osalpekar opened the pull request:
Summary:
Fixes #33480

This makes the `dist_autograd.backward` and `dist_optimizer.step` APIs functional by making the user explicitly pass in the `context_id` as opposed to relying on the potentially confusing `thread_local context_id`.

This diff incorporates these API changes and all places where these functions are called.

More concretely, this code:

```
with dist_autograd.context():
    # Forward pass.
    dist_autograd.backward([loss.sum()])
    dist_optim.step()
```

should now be written as follows:

```
with dist_autograd.context() as context_id:
    # Forward pass.
    dist_autograd.backward(context_id, [loss.sum()])
    dist_optim.step(context_id)
```

Test Plan: Ensuring all existing dist_autograd and dist_optimizer tests pass with the new API. Also added a new test case for input checking.

Differential Revision: D20011710

pritamdamania87 commented on torch/csrc/distributed/autograd/init.cpp:
```diff
@@ -152,8 +161,9 @@ Example::
     >> with dist_autograd.context() as context_id:
```
Update the "Arguments" section of the doc as well

pritamdamania87 commented on torch/distributed/optim/optimizer.py:
```diff
@@ -122,7 +122,7 @@ def __init__(self, optimizer_class, params_rref, *args, **kwargs):
-    def step(self):
+    def step(self, context_id):
         """
         Performs a single optimization step.
```
Add an Arguments section for this doc and mention context_id

pritamdamania87 reviewed (approved):
Nice stuff! Please update the docs as I've mentioned in my comments inline.

rohan-varma commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -297,10 +297,14 @@ std::shared_ptr<rpc::FutureMessage> DistEngine::executeSendFunctionAsync(
+    int64_t contextId,
+    const variable_list& roots,
+    bool retainGraph) {
   // Get the current context, if exists. This will throw if we don't have a
```
Nit: Update this comment?

osalpekar committed:
[Dist Autograd] Functional API for Dist Autograd and Dist Optimizer (#33711)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/33711

Fixed #33480

This makes `dist_autograd.backward` and `dist_optimizer.step` functional by making the user explicitly pass in the `context_id` as opposed to relying on the confusing thread_local context_id.

This diff incorporates these API changes and all places where these functions are called.

More concretely, this code:

```
with dist_autograd.context():
    # Forward pass.
    dist_autograd.backward([loss.sum()])
    dist_optim.step()
```

should now be written as follows:

```
with dist_autograd.context() as context_id:
    # Forward pass.
    dist_autograd.backward(context_id, [loss.sum()])
    dist_optim.step(context_id)
```

Test Plan: Ensuring all existing dist_autograd and dist_optimizer tests pass with the new API. Also added a new test case for input checking.

Differential Revision: D20011710

fbshipit-source-id: 368c801da31de86a1254a4707c706c60d2fadc26
